File: commands/src/main/java/com/github/netomi/bat/dump/InstructionPrinter.java
/*
 *  Copyright (c) 2020 Thomas Neidhart.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */
package com.github.netomi.bat.dump;

import com.github.netomi.bat.dexfile.*;
import com.github.netomi.bat.dexfile.instruction.*;
import com.github.netomi.bat.dexfile.visitor.InstructionVisitor;
import com.github.netomi.bat.util.Primitives;

import static com.github.netomi.bat.dexfile.instruction.DexOpCode.GOTO_32;

class      InstructionPrinter
implements InstructionVisitor
{
    private final BufferedPrinter printer;

    public InstructionPrinter(BufferedPrinter printer) {
        this.printer = printer;
    }

    @Override
    public void visitAnyInstruction(DexFile dexFile, ClassDef classDef, EncodedMethod method, Code code, int offset, DexInstruction instruction) {
        printGeneric(instruction);
    }

    @Override
    public void visitArithmeticInstruction(DexFile dexFile, ClassDef classDef, EncodedMethod method, Code code, int offset, ArithmeticInstruction instruction) {
        printGeneric(instruction);

        int literal = instruction.getLiteral();
        if (instruction.containsLiteral())  {
            printer.print(", ");
            printer.print("#int ");
            printer.print(Integer.toString(literal));
            printer.print(" // #");

            switch (instruction.getOpcode().getFormat()) {
                case FORMAT_22s:
                    printer.print(Primitives.asHexValue((short) literal));
                    break;

                case FORMAT_22b:
                    printer.print(Primitives.asHexValue((byte) literal));
                    break;
            }
        }
    }

    @Override
    public void visitBranchInstruction(DexFile dexFile, ClassDef classDef, EncodedMethod method, Code code, int offset, BranchInstruction instruction) {
        printGeneric(instruction);

        if (instruction.registers.length > 0) {
            printer.print(", ");
        } else {
            printer.print(" ");
        }

        if (instruction.getOpcode() == GOTO_32) {
            printer.print("#");
            if (instruction.getBranchOffset() < 0) {
                printer.print("-");
                printer.print(Primitives.asHexValue(-instruction.getBranchOffset(), 8));
            }
            else {
                printer.print(Primitives.asHexValue(instruction.getBranchOffset(), 8));
            }
        } else {
            printer.print(Primitives.asHexValue(offset + instruction.getBranchOffset(), 4));
            printer.print(" // ");

            if (instruction.getBranchOffset() < 0) {
                printer.print("-");
                printer.print(Primitives.asHexValue(-instruction.getBranchOffset(), 4));
            }
            else {
                printer.print("+");
                printer.print(Primitives.asHexValue(instruction.getBranchOffset(), 4));
            }
        }
    }

    @Override
    public void visitFieldInstruction(DexFile dexFile, ClassDef classDef, EncodedMethod method, Code code, int offset, FieldInstruction instruction) {
        printGeneric(instruction);

        printer.print(", ");

        FieldID fieldID = instruction.getField(dexFile);

        printer.print(fieldID.getClassType(dexFile));
        printer.print(".");
        printer.print(fieldID.getName(dexFile));
        printer.print(":");
        printer.print(fieldID.getType(dexFile));

        printer.print(" // field@");
        printer.print(Primitives.asHexValue(instruction.getFieldIndex(), 4));
    }

    @Override
    public void visitLiteralInstruction(DexFile dexFile, ClassDef classDef, EncodedMethod method, Code code, int offset, LiteralInstruction instruction) {
        printGeneric(instruction);

        printer.print(", ");

        long value = instruction.getValue();

        switch (instruction.getOpcode().getFormat()) {
            case FORMAT_11n:
            case FORMAT_22b:
                printer.print(String.format("#int %d // #%x", value, (byte) value));
                break;

            case FORMAT_21h:
                // The printed format varies a bit based on the actual opcode.
                if (instruction.getOpcode() == DexOpCode.CONST_HIGH16) {
                    short v = (short) (value >> 16);
                    printer.print(String.format("#int %d // #%x", value, v));
                } else {
                    short v = (short) (value >> 48);
                    printer.print(String.format("#long %d // #%x", value, v));
                }
                break;

            case FORMAT_21s:
            case FORMAT_22s:
                printer.print(String.format("#int %d // #%x", value, (short) value));
                break;

            case FORMAT_31i:
                printer.print(new PrintfFormat("#float %g // #%08x").sprintf(new Object[] { Float.intBitsToFloat((int) value), value }));
                break;

            case FORMAT_51l:
                printer.print(new PrintfFormat("#double %g // #%016lx").sprintf(new Object[] { Double.longBitsToDouble(value), value }));
                break;
        }
    }

    @Override
    public void visitMethodInstruction(DexFile dexFile, ClassDef classDef, EncodedMethod method, Code code, int offset, MethodInstruction instruction) {
        printer.print(instruction.getMnemonic());

        if (instruction.registers.length > 0) {
            printer.print(" ");
            printer.print("{");
            for (int idx = 0; idx < instruction.registers.length; idx++) {
                if (idx > 0) {
                    printer.print(", ");
                }
                printer.print("v");
                printer.print(Integer.toString(instruction.registers[idx]));
            }
            printer.print("}");
        } else {
            printer.print("{}");
        }

        printer.print(", ");

        MethodID methodID = instruction.getMethod(dexFile);

        printer.print(methodID.getClassType(dexFile));
        printer.print(".");
        printer.print(methodID.getName(dexFile));
        printer.print(":");
        printer.print(methodID.getProtoID(dexFile).getDescriptor(dexFile));

        printer.print(" // method@");
        printer.print(Primitives.asHexValue(instruction.getMethodIndex(), 4));
    }

    @Override
    public void visitPayloadInstruction(DexFile dexFile, ClassDef classDef, EncodedMethod method, Code code, int offset, PayloadInstruction instruction) {
        printGeneric(instruction);

        if (instruction.registers.length > 0) {
            printer.print(", ");
        } else {
            printer.print(" ");
        }

        printer.print(Primitives.asHexValue(offset + instruction.getPayloadOffset(), 8));
        printer.print(" // ");

        if (instruction.getPayloadOffset() < 0) {
            printer.print("-");
            printer.print(Primitives.asHexValue(-instruction.getPayloadOffset(), 8));
        } else {
            printer.print("+");
            printer.print(Primitives.asHexValue(instruction.getPayloadOffset(), 8));
        }
    }

    @Override
    public void visitStringInstruction(DexFile dexFile, ClassDef classDef, EncodedMethod method, Code code, int offset, StringInstruction instruction) {
        printGeneric(instruction);

        printer.printAsMutf8(", \"" + instruction.getString(dexFile) + "\" // string@", false);

        if (instruction.getOpcode() == DexOpCode.CONST_STRING) {
            printer.print(Primitives.asHexValue(instruction.getStringIndex(), 4));
        } else {
            printer.print(Primitives.asHexValue(instruction.getStringIndex(), 8));
        }
    }

    @Override
    public void visitTypeInstruction(DexFile dexFile, ClassDef classDef, EncodedMethod method, Code code, int offset, TypeInstruction instruction) {
        printGeneric(instruction);

        printer.print(", ");

        TypeID typeID = instruction.getTypeID(dexFile);
        printer.print(typeID.getType(dexFile));

        printer.print(" // type@");
        printer.print(Primitives.asHexValue(instruction.getTypeIndex(), 4));
    }

    @Override
    public void visitArrayTypeInstruction(DexFile dexFile, ClassDef classDef, EncodedMethod method, Code code, int offset, ArrayTypeInstruction instruction) {
        printer.print(instruction.getMnemonic());

        if (instruction.registers.length > 0) {
            printer.print(" ");
            printer.print("{");
            for (int idx = 0; idx < instruction.registers.length; idx++) {
                if (idx > 0) {
                    printer.print(", ");
                }
                printer.print("v");
                printer.print(Integer.toString(instruction.registers[idx]));
            }
            printer.print("}");
        } else {
            printer.print("{}");
        }

        printer.print(", ");

        TypeID typeID = instruction.getTypeID(dexFile);
        printer.print(typeID.getType(dexFile));

        printer.print(" // type@");
        printer.print(Primitives.asHexValue(instruction.getTypeIndex(), 4));
    }

    @Override
    public void visitFillArrayPayload(DexFile dexFile, ClassDef classDef, EncodedMethod method, Code code, int offset, FillArrayPayload payload) {
        printer.print(payload.toString());
    }

    @Override
    public void visitPackedSwitchPayload(DexFile dexFile, ClassDef classDef, EncodedMethod method, Code code, int offset, PackedSwitchPayload payload) {
        printer.print(payload.toString());
    }

    @Override
    public void visitSparseSwitchPayload(DexFile dexFile, ClassDef classDef, EncodedMethod method, Code code, int offset, SparseSwitchPayload payload) {
        printer.print(payload.toString());
    }

    private void printGeneric(DexInstruction instruction) {
        printer.print(instruction.getMnemonic());

        if (instruction.getOpcode() == DexOpCode.NOP) {
            printer.print(" // spacer");
        }

        if (instruction.registers.length > 0) {
            printer.print(" ");
            for (int idx = 0; idx < instruction.registers.length; idx++) {
                if (idx > 0) {
                    printer.print(", ");
                }
                printer.print("v");
                printer.print(Integer.toString(instruction.registers[idx]));
            }
        }
    }
}

File: dexdump/src/main/kotlin/com/github/netomi/bat/dexdump/InstructionPrinter.kt
/*
 *  Copyright (c) 2020 Thomas Neidhart.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */
package com.github.netomi.bat.dexdump

import com.github.netomi.bat.dexfile.ClassDef
import com.github.netomi.bat.dexfile.Code
import com.github.netomi.bat.dexfile.DexFile
import com.github.netomi.bat.dexfile.EncodedMethod
import com.github.netomi.bat.dexfile.instruction.*
import com.github.netomi.bat.dexfile.instruction.visitor.InstructionVisitor
import com.github.netomi.bat.util.toHexString
import com.github.netomi.bat.util.toSignedHexString
import java.lang.Double.longBitsToDouble
import java.lang.Float.intBitsToFloat

internal class InstructionPrinter(private val printer: Mutf8Printer) : InstructionVisitor {

    override fun visitAnyInstruction(dexFile: DexFile, classDef: ClassDef, method: EncodedMethod, code: Code, offset: Int, instruction: DexInstruction) {
        printGeneric(instruction)
    }

    override fun visitArithmeticLiteralInstruction(dexFile: DexFile, classDef: ClassDef, method: EncodedMethod, code: Code, offset: Int, instruction: ArithmeticLiteralInstruction) {
        printGeneric(instruction)

        val literal = instruction.literal
        printer.print(", #int $literal // #")

        when (instruction.opCode.format) {
            InstructionFormat.FORMAT_22s -> printer.print(toHexString(literal.toShort()))
            InstructionFormat.FORMAT_22b -> printer.print(toHexString(literal.toByte()))
            else -> error("unexpected format ${instruction.opCode.format} for arithmetic literal instruction")
        }
    }

    override fun visitBranchInstruction(dexFile: DexFile, classDef: ClassDef, method: EncodedMethod, code: Code, offset: Int, instruction: BranchInstruction) {
        printGeneric(instruction)

        if (instruction.registers.isNotEmpty()) {
            printer.print(", ")
        } else {
            printer.print(" ")
        }

        if (instruction.opCode == DexOpCode.GOTO_32) {
            printer.print("#")
            if (instruction.branchOffset < 0) {
                printer.print("-")
                printer.print(toHexString(-instruction.branchOffset, 8))
            } else {
                printer.print(toHexString(instruction.branchOffset, 8))
            }
        } else {
            printer.print(toHexString(offset + instruction.branchOffset, 4))
            printer.print(" // ")
            printer.print(toSignedHexString(instruction.branchOffset, 4))
        }
    }

    override fun visitCallSiteInstruction(dexFile: DexFile, classDef: ClassDef, method: EncodedMethod, code: Code, offset: Int, instruction: CallSiteInstruction) {
        printer.print(instruction.mnemonic)
        printRegisters(instruction, true)
        printer.print(", call_site@${toHexString(instruction.callSiteIndex, 4)}")
    }

    override fun visitFieldInstruction(dexFile: DexFile, classDef: ClassDef, method: EncodedMethod, code: Code, offset: Int, instruction: FieldInstruction) {
        printGeneric(instruction)
        val fieldID = instruction.getField(dexFile)
        printer.print(", ${fieldID.getClassType(dexFile)}.${fieldID.getName(dexFile)}:${fieldID.getType(dexFile)}")
        printer.print(" // field@${toHexString(instruction.fieldIndex, 4)}")
    }

    override fun visitLiteralInstruction(dexFile: DexFile, classDef: ClassDef, method: EncodedMethod, code: Code, offset: Int, instruction: LiteralInstruction) {
        printGeneric(instruction)
        printer.print(", ")

        val value = instruction.value
        when (instruction.opCode.format) {
            InstructionFormat.FORMAT_11n,
            InstructionFormat.FORMAT_22b -> printer.print("#int %d // #%x".format(value, value.toByte()))

            InstructionFormat.FORMAT_21h -> {
                // The printed format varies a bit based on the actual opcode.
                if (instruction.opCode == DexOpCode.CONST_HIGH16) {
                    val v = (value shr 16).toShort()
                    printer.print("#int %d // #%x".format(value, v))
                } else {
                    val v = (value shr 48).toShort()
                    printer.print("#long %d // #%x".format(value, v))
                }
            }

            InstructionFormat.FORMAT_21s,
            InstructionFormat.FORMAT_22s -> printer.print("#int %d // #%x".format(value, value.toShort()))

            InstructionFormat.FORMAT_31i -> printer.print("#float %g // #%08x".format(intBitsToFloat(value.toInt()), value.toInt()))

            InstructionFormat.FORMAT_51l -> printer.print("#double %g // #%016x".format(longBitsToDouble(value), value))

            else -> error("unexpected format ${instruction.opCode.format} for literal instruction")
        }
    }

    override fun visitAnyMethodInstruction(dexFile: DexFile, classDef: ClassDef, method: EncodedMethod, code: Code, offset: Int, instruction: MethodInstruction) {
        printer.print(instruction.mnemonic)
        printRegisters(instruction, true)
        printer.print(", ")

        val polymorphicMethodInstruction = if (instruction is MethodProtoInstruction) instruction else null
        val methodID = instruction.getMethodID(dexFile)

        printer.print("${methodID.getClassType(dexFile)}.${methodID.getName(dexFile)}:${methodID.getProtoID(dexFile).getDescriptor(dexFile)}")

        polymorphicMethodInstruction?.apply {
            printer.print(", " + this.getProtoID(dexFile).getDescriptor(dexFile))
        }

        printer.print(" // method@")
        printer.print(toHexString(instruction.methodIndex, 4))

        polymorphicMethodInstruction?.apply {
            printer.print(", proto@")
            printer.print(toHexString(polymorphicMethodInstruction.protoIndex, 4))
        }
    }

    override fun visitMethodHandleRefInstruction(dexFile: DexFile, classDef: ClassDef, method: EncodedMethod, code: Code, offset: Int, instruction: MethodHandleRefInstruction) {
        printGeneric(instruction)
        printer.print(", method_handle@${toHexString(instruction.methodHandleIndex, 4)}")
    }

    override fun visitMethodTypeRefInstruction(dexFile: DexFile, classDef: ClassDef, method: EncodedMethod, code: Code, offset: Int, instruction: MethodTypeRefInstruction) {
        printGeneric(instruction)
        printer.print(", ")
        printer.print(instruction.getProtoID(dexFile).getDescriptor(dexFile))
        printer.print(" // proto@")
        printer.print(toHexString(instruction.protoIndex, 4))
    }

    override fun visitAnyPayloadInstruction(dexFile: DexFile, classDef: ClassDef, method: EncodedMethod, code: Code, offset: Int, instruction: PayloadInstruction) {
        printGeneric(instruction)
        if (instruction.registers.isNotEmpty()) {
            printer.print(", ")
        } else {
            printer.print(" ")
        }
        printer.print(toHexString(offset + instruction.payloadOffset, 8))
        printer.print(" // ")
        printer.print(toSignedHexString(instruction.payloadOffset, 8))
    }

    override fun visitStringInstruction(dexFile: DexFile, classDef: ClassDef, method: EncodedMethod, code: Code, offset: Int, instruction: StringInstruction) {
        printGeneric(instruction)

        printer.print(", \"")
        printer.printAsMutf8(instruction.getString(dexFile), false)
        printer.print("\" // string@")

        if (instruction.opCode == DexOpCode.CONST_STRING) {
            printer.print(toHexString(instruction.stringIndex, 4))
        } else {
            printer.print(toHexString(instruction.stringIndex, 8))
        }
    }

    override fun visitTypeInstruction(dexFile: DexFile, classDef: ClassDef, method: EncodedMethod, code: Code, offset: Int, instruction: TypeInstruction) {
        printGeneric(instruction)
        printer.print(", ")
        val typeID = instruction.getTypeID(dexFile)
        printer.print(typeID.getType(dexFile))
        printer.print(" // type@")
        printer.print(toHexString(instruction.typeIndex, 4))
    }

    override fun visitArrayTypeInstruction(dexFile: DexFile, classDef: ClassDef, method: EncodedMethod, code: Code, offset: Int, instruction: ArrayTypeInstruction) {
        printer.print(instruction.mnemonic)
        printRegisters(instruction, true)
        printer.print(", ")
        val typeID = instruction.getTypeID(dexFile)
        printer.print(typeID.getType(dexFile))
        printer.print(" // type@")
        printer.print(toHexString(instruction.typeIndex, 4))
    }

    override fun visitAnyPayload(dexFile: DexFile, classDef: ClassDef, method: EncodedMethod, code: Code, offset: Int, payload: Payload) {
        printer.print(payload.toString())
    }

    private fun printGeneric(instruction: DexInstruction) {
        printer.print(instruction.mnemonic)
        if (instruction.opCode == DexOpCode.NOP) {
            printer.print(" // spacer")
        }
        printRegisters(instruction, false)
    }

    private fun printRegisters(instruction: DexInstruction, useBrackets: Boolean) {
        if (useBrackets) {
            if (instruction.registers.isNotEmpty()) {
                val registers = instruction.registers.indices.joinToString(", ", "{", "}", transform = { idx -> "v${instruction.registers[idx]}"})
                printer.print(" $registers")
            } else {
                printer.print(" {}")
            }
        } else {
            if (instruction.registers.isNotEmpty()) {
                val registers = instruction.registers.indices.joinToString(", ", transform = { idx -> "v${instruction.registers[idx]}"})
                printer.print(" $registers")
            }
        }
    }
}